### DISCUSSION TITLE
Some (required_without_all and different) form field validation rules seems to be not respected

### DISCUSSION THREAD

**Discussion by studiowizjo** [filamentphp/filament#2585] [Help] [answered]
As it stays in topic, here are some cases:

**Case 1: "required_without_all" issue**:
```php
Select::make('project_id')
  ->label('Project')
  ->options(Project::pluck('name', 'id'))
  ->searchable()
  ->rules(['required_without_all:company_id,task_id']),
Select::make('company_id')
  ->label('Company')
  ->options(Company::pluck('name', 'id'))
  ->searchable()
  ->rules(['required_without_all:project_id,task_id']),
TextInput::make('task_id')
  ->label('Task')
  ->prefix('#')
  ->numeric()
  ->rules(['required_without_all:company_id,project_id']),
```

Expected behaviour: At least one of fields should be required, but more than one can be completed also
Observed behaviour: One of the fields is populated, but two others still shows error

**Case 2: "different" issue**:
```php
TextInput::make('hour_from')
  ->label('From hour')
  ->type('time')
  ->default(date('H:i'))
  ->rules('different:hour_to')
  ->required(),
TextInput::make('hour_to')
  ->label('To hour')
  ->type('time')
  ->default(date('H:i'))
  ->rules('different:hour_from')
  ->required(),
```

Expected behaviour: Typed hours should be different
Observed behaviour: There is no error, event if the value in fields is the same

**Comment by Z3d0X** [comment, marked as answer]
https://filamentphp.com/docs/2.x/forms/validation#getting-started

For 2nd case you can use
`->different('field_name')`

For 1st case
As far as I'm aware there isn't a `requiredWithOut()` method (This would be good addition though)

So for now you can do something like
`->required(fn ($get) => blank($get('without_field')) && blank($get('another_field)) )`

**Reply by studiowizjo** [reply]
Thank you. Everything works as expected. I didn`t came up with idea of using closure for **required()** method like this.

That solves topic, but still I think that something may be a bit wrong, because [documentation about rules](https://filamentphp.com/docs/2.x/forms/validation#other-rules) says:

> You may add other validation rules to any field using the rules() method
> You may use any custom validation rules as you would do in [Laravel](https://laravel.com/docs/validation#custom-validation-rules):

and gives example:

```php
TextInput::make('slug')->rules(['alpha_dash'])
```

so, according to this rules I provided should also work.

Anyway, thanks again =).

**Reply by Z3d0X** [reply]
Docs also state this

> Beware that some validations rely on the field name and therefore won't work when passed via ->rule()/->rules(). Use the dedicated validation methods whenever you can

**Reply by danharrin** [reply]
In the next version, every single validation rule should now be present

**Comment by danharrin** [comment]
The reason for this is because in the admin panel, fields are stored in a $data array, so `data.company_id` or `data.task_id` would have worked fine.

I would appreciate PR additions of new validation methods for these rules so others don't get caught out.

**Comment by ManojKiranA** [comment]
@danharrin what would you suggest for preparing the rules with prefix. May be we can loop through each rules and chech for validation rules such as `required_without_all`, `date_after` , etc  and add prefix `data.`

Is there any way to get the statePath of the form or we can use the `data.` directly.

Let me know what you think

**Reply by danharrin** [reply]
It's something like this:

https://github.com/filamentphp/filament/blob/2.x/packages/forms/src/Components/Concerns/CanBeValidated.php#L250-L260

**Comment by Z3d0X** [comment]
https://github.com/filamentphp/filament/pull/3682
